colflow: create separate memory accounts for streaming components

yuzefovich opened the pull request:
**colexec: move Allocator in hash router output under the mutex**

The Allocator object is shared among the output of a hash router, so
access to it needs to be synchronized. It just so happens that it was
already the case (we acquire the lock in the beginning of 'addBatch' and
defer its release), and this commit moves the Allocator into 'mu' struct
to make it explicit.

Release note: None

**mon: minor tweaks to BytesMonitor**

It increases the number of lines from the stack trace when verbose
logging is enabled. Also, it correctly logs the amount of bytes requested
to release is greater than currently allocated.

Release note: None

**colflow: create separate memory accounts for streaming components**

Previously, all non-buffering vectorized components shared the same
memory account. However, those components can run concurrently (for
example, multiple Outboxes on the same node), and we could get into
a race situation because the access to the single memory account wasn't
synchronized (although different Allocator objects are used, they shared
the same account). Now this is fixed by creating separate memory
accounts for all streaming components.

Fixes: #41322.
Fixes: #42722.

Release note: None

yuzefovich committed:
colexec: move Allocator in hash router output under the mutex

The Allocator object is shared among the output of a hash router, so
access to it needs to be synchronized. It just so happens that it was
already the case (we acquire the lock in the beginning of 'addBatch' and
defer its release), and this commit moves the Allocator into 'mu' struct
to make it explicit.

Release note: None

yuzefovich committed:
mon: minor tweaks to BytesMonitor

It increases the number of lines from the stack trace when verbose
logging is enabled. Also, it correctly logs the amount of bytes requested
to release is greater than currently allocated.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/42756)
<!-- Reviewable:end -->

yuzefovich commented:
Unfortunately, this doesn't fix the memory accounting issue - `allocator` happened to be under the "implicit" lock, so my theory was wrong. I'll put off the investigation of this issue for now :/

yuzefovich committed:
colflow: create separate memory accounts for streaming components

Previously, all non-buffering vectorized components shared the same
memory account. However, those components can run concurrently (for
example, multiple Outboxes on the same node), and we could get into
a race situation because the access to the single memory account wasn't
synchronized (although different Allocator objects are used, they shared
the same account). Now this is fixed by creating separate memory
accounts for all streaming components.

Release note: None

yuzefovich commented:
I pushed the changes I mentioned about the streaming memory accounts and have been stressing the test for 10 minutes now (no failures, so I'm feeling optimistic).

I'm not sure whether we want to close #41322 because the first failure in there happened month and a half ago and is unrelated to the problems this PR is fixing. Should we close that issue anyway since there was no other instances of the same failure?

yuzefovich commented:
No repro in 45 minutes (8000 runs), so I think this is it. :yay:

asubiotto reviewed (commented):
> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42756)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/util/mon/bytes_usage.go, line 635 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42756#-Lv5K_ufCr8WPJ_UP5rJ:-Lv5K_ufCr8WPJ_UP5rK:b-ahc95r) ([raw file](https://github.com/cockroachdb/cockroach/blob/0152b81929b3cf8c251fbbe4fa8036cb27a51b9a/pkg/util/mon/bytes_usage.go#L635)):*
> ```Go
> 	defer mm.mu.Unlock()
> 	if mm.mu.curAllocated < sz {
> 		sz = mm.mu.curAllocated
> ```

I think this is so that we don't get a negative `mm.mu.curAllocated` below so we can free bytes correctly when we don't panic (i.e. in production). I think this should probably still happen but *after* the log message.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42756)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/util/mon/bytes_usage.go, line 635 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42756#-Lv5K_ufCr8WPJ_UP5rJ:-Lv6X6jLBPUAR4tL-ilj:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/0152b81929b3cf8c251fbbe4fa8036cb27a51b9a/pkg/util/mon/bytes_usage.go#L635)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think this is so that we don't get a negative `mm.mu.curAllocated` below so we can free bytes correctly when we don't panic (i.e. in production). I think this should probably still happen but *after* the log message.
</blockquote></details>

Good point, done.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42756#-:-Lv6Xzjn2Dz8aVVWW53w:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+
